trawl.go

```diff
@@ -31,81 +31,54 @@ func New(iface net.Interface) (i *Interface, err error) {
 	ipv4, ipv6 := extractAddrs(addrs)
 
 	// if we have an IPv6 only interface
-	if len(ipv4) == 0 {
+	if ipv4 == nil {
 		return &Interface{
 			Name:     iface.Name,
-			IPv6Addr: ipv6,
+			IPv6Addr: ipv6.String(),
 		}, nil
 	}
 
-	// get IPv4 address & dotted decimal mask
-	ipv4Split := strings.Split(ipv4, "/")
-	ipv4addr := ipv4Split[0]
-	ipv4Cidr := ipv4Split[1]
-	ipv4Mask, err := toDottedDec(ipv4Cidr)
-	if err != nil {
-		return i, err
-	}
-
-	// get IPv4 network
-	_, ipnet, err := net.ParseCIDR(ipv4)
-	if err != nil {
-		return i, err
-	}
-
 	return &Interface{
 		HardwareAddr: iface.HardwareAddr.String(),
-		IPv4Addr:     ipv4addr,
-		IPv4Mask:     ipv4Mask,
-		IPv4Network:  ipnet.String(),
-		IPv6Addr:     ipv6,
+		IPv4Addr:     ipv4.IP.String(),
+		IPv4Mask:     toDottedDec(ipv4.Mask),
+		IPv4Network:  maskedIPString(ipv4),
+		IPv6Addr:     ipv6.String(),
 		MTU:          iface.MTU,
 		Name:         iface.Name,
 	}, nil
 }
 
-func extractAddrs(addrs []net.Addr) (ipv4 string, ipv6 string) {
+func extractAddrs(addrs []net.Addr) (ipv4, ipv6 *net.IPNet) {
 	for _, addr := range addrs {
-		if a := addr.String(); strings.Contains(a, ":") {
-			ipv6 = a
-		}
-		if a := addr.String(); strings.Contains(a, ".") {
-			ipv4 = a
+		switch ipnet := addr.(type) {
+		case *net.IPNet:
+			if ip := ipnet.IP.To4(); ip != nil {
+				ipv4 = ipnet
+			} else {
+				ipv6 = ipnet
```

Comment:
This blows up if there's no IPv6 address, when the `String()` method is called on line 46 
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x485696]

goroutine 1 [running]:
panic(0x603ea0, 0xc42000c1f0)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
net.networkNumberAndMask(0x0, 0x3, 0xc42000d360, 0xd, 0x4, 0x4, 0x3)
	/usr/local/go/src/net/ip.go:443 +0x26
net.(*IPNet).String(0x0, 0xc42000d360, 0xd)
	/usr/local/go/src/net/ip.go:493 +0x40
main.New(0x3, 0x5dc, 0xc42000d1a0, 0x7, 0xc4200d49f0, 0x6, 0xf10, 0x13, 0xc4200180e0, 0x0, ...)
	/home/rob/code/go/src/github.com/robphoenix/trawl/trawl.go:47 +0x17c
main.main()
	/home/rob/code/go/src/github.com/robphoenix/trawl/main.go:62 +0x2ca
```